graphix/parameter.py

```diff
@@ -0,0 +1,258 @@
+from typing import TYPE_CHECKING
+
+import pydantic_core
+
+if TYPE_CHECKING:
+    import pydantic
+
+
+class Expression(ABC):
+    """Expression with parameters."""
+
+    @abstractmethod
+    def __mul__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __rmul__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __add__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __radd__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __sub__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __rsub__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __pow__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __rpow__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __neg__(self) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __truediv__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __rtruediv__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def __mod__(self, other) -> ExpressionOperatorResult: ...
+
+    @abstractmethod
+    def sin(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def cos(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def tan(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def arcsin(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def arccos(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def arctan(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def exp(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def log(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def conjugate(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def sqrt(self) -> ExpressionOrNumber: ...
+
+    @abstractmethod
+    def __repr__(self) -> str: ...
+
+    @abstractmethod
+    def __str__(self) -> str: ...
+
+    @abstractmethod
+    def subs(self, variable: Parameter, value: ExpressionOrNumber) -> ExpressionOrNumber: ...
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source_type: typing.Any, handler: pydantic.GetCoreSchemaHandler
+    ) -> pydantic_core.CoreSchema:
+        def check_expression(obj) -> Expression:
+            if not isinstance(obj, Expression):
+                raise ValueError("Expression expected")
+            return obj
+
+        return pydantic_core.core_schema.no_info_plain_validator_function(function=check_expression)
+
+
+class PlaceholderOperationError(ValueError):
+    def __init__(self):
+        super().__init__(
+            "Placeholder angles do not support any form of computation before substitution. Either use `subs` with an actual value before the computation, or use a symbolic parameter implementation, such that https://github.com/TeamGraphix/graphix-symbolic ."
```

Comment:
typo: such **as** https://github.com/TeamGraphix/graphix-symbolic

Reply:
Fixed in 49e63df .